pkg/cmd/roachtest/testselector/selector.go

```diff
@@ -0,0 +1,205 @@
+	return &SelectTestsReq{
+		ForPastDays:          defaultForPastDays,
+		FirstRunOn:           defaultFirstRunOn,
+		LastRunOn:            defaultLastRunOn,
+		SelectFromSuccessPct: selectFromSuccessPct,
+		Cloud:                cloud,
+		Suite:                suite,
+	}
+}
+
+// CategoriseTests returns the tests categorized based on the snowflake query
+// The tests are Selected by selector.go based on certain criteria:
+// 1. the number of time a test has been successfully running
+// 2. the test is new
+// 3. the test has not been run for a while
+// 4. a subset of the successful tests based on SelectTestReq.SelectFromSuccessPct
+// It returns all the tests. The selected tests have the value TestDetails.Selected as true
+func CategoriseTests(ctx context.Context, req *SelectTestsReq) ([]*TestDetails, error) {
+	db, err := getConnect(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = db.Close() }()
+	statement, err := db.Prepare(preparedQuery)
+	if err != nil {
+		return nil, err
+	}
+	// add the parameters in sequence
+	rows, err := statement.QueryContext(ctx,
+		req.ForPastDays*-1, fmt.Sprintf("%%%s - %s%%", suites[req.Suite], req.Cloud),
+		req.FirstRunOn*-1, req.LastRunOn*-1)
+	if err != nil {
+		return nil, err
+	}
+	// All the column headers
+	colHeaders, err := rows.Columns()
+	if err != nil {
+		return nil, err
+	}
+	// Will be used to read data while iterating rows.
+	colPointers := make([]interface{}, len(colHeaders))
+	colContainer := make([]string, len(colHeaders))
+	for i := range colPointers {
+		colPointers[i] = &colContainer[i]
+	}
+	// allTestDetails are all the tests that are obtained from snowflake query
+	allTestDetails := make([]*TestDetails, 0)
+	// skipped tests are maintained separately
+	// this helps in considering them for running based on further select criteria like selectFromSuccessPct
+	skippedTests := make([]*TestDetails, 0)
+	for rows.Next() {
+		_ = rows.Scan(colPointers...)
+		testInfos := make([]string, len(colContainer))
+		copy(testInfos, colContainer)
+		// csv columns:
+		// 0. TEST_NAME
+		// 1. SELECTED (yes/no)
+		// 2. AVG_DURATION
+		// 3. TOTAL_RUNS
+		testDetails := &TestDetails{
+			Name:                testInfos[0],
+			Selected:            testInfos[1] != "no",
+			AvgDurationInMillis: getDuration(testInfos[2]),
+			TotalRuns:           getTotalRuns(testInfos[3]),
+		}
+		if !testDetails.Selected {
+			// skipped based on query
+			skippedTests = append(skippedTests, testDetails)
+		} else {
+			// Selected for running
+			allTestDetails = append(allTestDetails, testDetails)
+		}
+	}
+	if req.SelectFromSuccessPct > 0 && len(skippedTests) > 0 {
+		// need to select some tests from th success list (skipped tests)
+		numberOfTestsToSelect := len(skippedTests) * req.SelectFromSuccessPct / 100
+		// the tests are sorted by the number of runs. So, simply iterate over teh list
+		// and select the first count of "numberOfTestsToSelect"
+		for i := range skippedTests {
```

Comment:
Nit: we don't need to gate this logic necessarily. If either `SelectFromSuccessPct` or `len(skippedTests)` is zero, `numberOfTestsToSelect` will also be zero and the loop will trivially not update any tests. This is already the case anyway if `numberOfTestsToSelect` rounds down to zero due to the floating point coercion to int.

I also personally feel the loop is a little clearer as defined below, but maybe it's just me (i.e., feel free to ignore if you disagree):

```go
for i := 0; i < numberOfTestsToSelect; i++ {
    skippedTests[i].Selected = true
}
```